onchaind/onchain.c

```diff
@@ -97,99 +100,105 @@ struct tracked_output {
 	struct resolution *resolved;
 };
 
-/* We use the same feerate for htlcs and commit transactions; we don't
- * record what it was, so we brute-force it. */
-struct {
-	u32 min, max;
-} feerate_range;
-
-static void init_feerate_range(u64 funding_satoshi,
-			       const struct bitcoin_tx *commit_tx)
-{
-	size_t i, max_untrimmed_htlcs;
-	u64 fee = funding_satoshi;
-
-	for (i = 0; i < tal_count(commit_tx->output); i++)
-		fee -= commit_tx->output[i].amount;
-
-	/* We don't know how many trimmed HTLCs there are, so they could
-	 * be making fee entirely. */
-	feerate_range.min = 0;
-
-	/* But we can estimate the maximum fee rate:
-	 *
-	 * fee = feerate_per_kw * (724 + 172 * num_untrimmed) / 1000;
-	 */
-	if (tal_count(commit_tx->output) < 2)
-		max_untrimmed_htlcs = 0;
-	else
-		max_untrimmed_htlcs = tal_count(commit_tx->output) - 2;
-
-	feerate_range.max = (fee + 999) * 1000
-		/ (724 + 172 * max_untrimmed_htlcs);
-
-	status_trace("Initial feerate %u to %u",
-		     feerate_range.min, feerate_range.max);
-}
-
-static void narrow_feerate_range(u64 fee, u32 multiplier)
-{
-	u32 min, max;
-
-	/* fee = feerate_per_kw * multiplier / 1000; */
-
-	max = (fee + 999) * 1000 / multiplier;
-	if (fee < 999)
-		min = 0;
-	else
-		min = (fee - 999) * 1000 / multiplier;
-
-	status_trace("Fee %"PRIu64" gives feerate min/max %u/%u",
-		     fee, min, max);
-	if (max < feerate_range.max)
- 		feerate_range.max = max;
-	if (min > feerate_range.min)
- 		feerate_range.min = min;
-	status_trace("Feerate now %u to %u",
-		     feerate_range.min, feerate_range.max);
-}
-
-/* We vary feerate until signature they offered matches: we're more
- * likely to be near max. */
-static bool grind_feerate(struct bitcoin_tx *commit_tx,
-			  const secp256k1_ecdsa_signature *remotesig,
-			  const u8 *wscript,
-			  u64 multiplier)
+/* We vary feerate until signature they offered matches. */
+static u64 grind_htlc_tx_fee(struct bitcoin_tx *tx,
+			     const secp256k1_ecdsa_signature *remotesig,
+			     const u8 *wscript,
+			     u64 multiplier)
 {
 	u64 prev_fee = UINT64_MAX;
-	u64 input_amount = *commit_tx->input[0].amount;
+	u64 input_amount = *tx->input[0].amount;
 
-	for (s64 i = feerate_range.max; i >= feerate_range.min; i--) {
+	for (u64 i = min_possible_feerate; i <= max_possible_feerate; i++) {
 		u64 fee = i * multiplier / 1000;
```

Comment:
I remember branch `bitcoind-vbytes-must-die-die-die` needed to add some complicated computations to match bitcoind computation of feerate, should we also use it here?  Or this is sufficient? How does e.g. lnd compute this?